### ISSUE TITLE
Splitting up layers into smaller components

### ISSUE THREAD

**Issue by bvanessen** [LLNL/lbann#76]
Right now we have - linear, non-linear, regularization, etc.
       - We could have each layer as one of the smaller components
       - Big monolithic layers - we are forced to keep around a big
       block of memory
       - Tim: pooling has to copy the data - incurs some non-trivial
       performance cost
       - Nikoli: batch-norm - layers overwrite linearity results -
       now batch norm has to copy and save the state

**Comment by bvanessen** [comment]
I want to leave this one until Monday.  There is a sequencing problem here and I would like to get the hoist to templates done before we start on this one.

**Comment by bvanessen** [comment]
Finished in the release-v0.93 branch. 59bc798bb8b6b3b7ceb43a186913d5fa40a9192c
